posts/2018-11-28-http4s-error-handling-mtl-2.md

```diff
@@ -0,0 +1,459 @@
+    - org.typelevel::cats-effect:1.1.0-M1
+    - org.http4s::http4s-blaze-server:0.20.0-M3
+    - org.http4s::http4s-circe:0.20.0-M3
+    - org.http4s::http4s-dsl:0.20.0-M3
+    - io.circe::circe-core:0.10.0
+    - io.circe::circe-generic:0.10.0
+    - co.fs2::fs2-core:1.0.0
+    - com.olegpy::meow-mtl:0.2.0
+    - com.chuusai::shapeless:2.3.3
+
+---
+
+This is a continuation of my [previous blog post](https://typelevel.org/blog/2018/08/25/http4s-error-handling-mtl.html). Make sure you have read that one before continuing here.
+
+I recently gave a 20 minutes talk on `classy optics` at the unconference of [Scale by the Bay](http://scale.bythebay.io/) where I also talked about this error handling technique and on my way back home I was still thinking of different ways of doing this. So, after some exploratory work, I came up with a few different alternatives.
+
+### Issues with first approach
+
+Something that made me cringe and that a few of my colleagues at work were not happy with was that the algebras had no association with the error type defined in `HttpErrorHandler[F, E]` so the type-safety was down to the programmer's discipline and in this case the compiler was not able to do much.
+
+When working with `EitherT[F, E, A]` or a bifunctor `IO[E, A]` we have a clear error type whereas by just relying on a single `F[A]` with a `MonadError[F, Throwable]` instance we lose this property. However, most of the time we only care about mapping a few business errors to Http responses if working on a REST API.
+
+#### Data types with error handling capabilities
+
+As of now I still prefer the latter for a number of reasons. Here's a comparison table of all the different approaches I can think of:
+
+|            | EitherT[IO, E, ?] | IO[E, A] | IO[A] | IO[A] + Classy Optics |
+|------------|-------------------|----------|-------|-----------------------|
+| Concise code style | ✘ | ✔ | ✔ | ✔ |
+| Polymorphic code | ✔ | ✘ | ✔ | ✔ |
+| Performance overhead | **2x** | ✔ | ✔ | ✔ |
+| Double error channel | ✔ | ✔ | ✘ | ✘ |
+
+With "Double error channel" I mean that we can signal failure in two different ways:
+
+- `EitherT[IO, E, ?]` can either fail with `Left` or with `IO.raiseError` (that can also be a caught exception).
+- `IO[E, A]` can either fail with an `E` or by throwing an exception when `A` was expected. This is called "unrecoverable" errors.
+
+And this is not a bad thing at all! It is like this by design and we can have the same property when writing polymorphic code using `cats-effect` while still keeping it simple. Here's one way:
+
+### Error Channel
+
+In the previous blog post we defined the algebras as a single trait. In this case we are going to try a different encoding but first we need to introduce an `ErrorChannel[F, E]` typeclass where the error type is a subtype of `Throwable` to be compatible with the error type of the `cats-effect` typeclasses:
+
+```tut:book:silent
+trait ErrorChannel[F[_], E <: Throwable] {
```

Comment:
Why can't we just use `ApplicativeError` instead of inventing a new typeclass isomorphic to `ApplicativeError`?  Are we trying to avoid ambiguity with `ApplicativeError[F, Throwable]` from `F`'s other error channel?

Reply:
Two reasons why:

1. We can't have `ApplicativeError[F, UserError]` when we already have `Sync[F]` in scope that brings in `ApplicativeError[F, Throwable]`.
2. If we were to use `Sync[F].raiseError` we lose the type-safety we were trying to gain. The idea is that if somewhere in the call-chain there's a failure with `Throwable` in `F` that should be considered unrecoverable. Whereas if you use the `ErrorChannel` to raise errors you can be sure to handle your custom errors defined as an ADT. And you cannot do `ErrorChannel[F, UserError].raise(new Exception("boom"))` either as that won't compile.